syntax = "proto3";

import "entity.proto";

package quran;

message PartsRequest {
    string edition_id = 1;
    string list = 2;
}

message AyahIdRequest {
    string id = 1;
    PartsRequest parts = 2;
}

message AyahNumberRequest {
    int32 number = 1;
    PartsRequest parts = 2;
}

message AyahSajdaRequest {
    bool sajda = 1;
    PartsRequest parts = 2;
}

message AyahResponse {
    AyahEntity ayah = 1;
    TranslationEntity translation = 2;
    SurahEntity surah = 3;
    EditionEntity edition = 4;
    AudioEntity arabic_audio = 5;
    AudioEntity translation_audio = 6;
    ImageEntity image = 7;
}

message AyahEntityResponse {
    int32 code = 1;
    string status = 2;
    AyahEntity data = 3;
}

message AyahSingleResponse {
    int32 code = 1;
    string status = 2;
    AyahResponse data = 3;
}

message AyahMultiResponse {
    int32 code = 1;
    string status = 2;
    repeated AyahResponse data = 3;
}

service Ayah {
    rpc CreateAyah (AyahEntity) returns (AyahEntityResponse) {}
    rpc FindAyahById (AyahIdRequest) returns (AyahSingleResponse) {}
    rpc FindAyahBySurahId (AyahIdRequest) returns (AyahMultiResponse) {}
    rpc FindAyahByNumber (AyahNumberRequest) returns (AyahSingleResponse) {}
    rpc FindAyahByNumberInSurah (AyahNumberRequest) returns (AyahSingleResponse) {}
    rpc FindAyahByJuz (AyahNumberRequest) returns (AyahMultiResponse) {}
    rpc FindAyahByManzil (AyahNumberRequest) returns (AyahMultiResponse) {}
    rpc FindAyahByRuku (AyahNumberRequest) returns (AyahMultiResponse) {}
    rpc FindAyahByHizbQuarter (AyahNumberRequest) returns (AyahMultiResponse) {}
    rpc FindAyahBySajda (AyahSajdaRequest) returns (AyahMultiResponse) {}
}